devel/fortio/fortioLib/periodic.go

```diff
@@ -0,0 +1,244 @@
+		}
+		wg.Wait()
+		for t := 0; t < r.NumThreads; t++ {
+			functionDuration.Transfer(fDs[t])
+			sleepTime.Transfer(sDs[t])
+		}
+	}
+	elapsed := time.Since(start)
+	actualQPS := float64(functionDuration.Count) / elapsed.Seconds()
+	fmt.Printf("Ended after %v : %d calls. qps=%.5g\n", elapsed, functionDuration.Count, actualQPS)
+	if useQPS {
+		percentNegative := 100. * float64(sleepTime.hdata[0]) / float64(sleepTime.Count)
+		if percentNegative > 3 {
+			sleepTime.FPrint(os.Stdout, "Aggregated Sleep Time", 50)
+			fmt.Printf("WARNING %.2f%% of sleep were falling behind\n", percentNegative)
+		} else {
+			if Verbose > 0 {
+				sleepTime.FPrint(os.Stdout, "Aggregated Sleep Time", 50)
+			} else {
+				sleepTime.Counter.FPrint(os.Stdout, "Sleep times")
+			}
+		}
+	}
+	functionDuration.FPrint(os.Stdout, "Aggregated Function Time", r.Percentiles[0])
+	for _, p := range r.Percentiles[1:] {
+		fmt.Printf("# target %g%% %.6g\n", p, functionDuration.CalcPercentile(p))
+	}
+}
+
+func runOne(t int, cF *Histogram, cS *Histogram, numCalls int64, start time.Time, r *periodicRunner) {
+	var i int64
+	endTime := start.Add(r.Duration)
+	tIDStr := fmt.Sprintf("T%03d", t)
+	perThreadQPS := r.QPS / float64(r.NumThreads)
+	useQPS := (perThreadQPS > 0)
+	verbose := r.Verbose
+	f := r.Function
+	for {
+		fStart := time.Now()
+		if fStart.After(endTime) {
+			if !useQPS {
+				// max speed test reached end:
+				break
+			}
+			// QPS mode:
+			// Do least 2 iterations, and the last one before bailing because of time
+			if (i >= 2) && (i != numCalls-1) {
+				log.Printf("%s warning only did %d out of %d calls before reaching %v", tIDStr, i, numCalls, r.Duration)
+				break
+			}
+		}
+		f(t)
+		cF.Record(time.Since(fStart).Seconds())
+		i++
+		// if using QPS / pre calc expected call # mode:
+		if useQPS {
+			if i >= numCalls {
+				break // expected exit for that mode
+			}
+			elapsed := time.Since(start)
+			// This next line is tricky - such as for 2s duration and 1qps there is 1
+			// sleep of 2s between the 2 calls and for 3qps in 1sec 2 sleep of 1/2s etc
+			targetElapsedInSec := (float64(i) + float64(i)/float64(numCalls-1)) / perThreadQPS
+			targetElapsedDuration := time.Duration(int64(targetElapsedInSec * 1e9))
+			sleepDuration := targetElapsedDuration - elapsed
+			if verbose > 3 {
+				log.Printf("%s target next dur %v - sleep %v", tIDStr, targetElapsedDuration, sleepDuration)
+			}
+			cS.Record(sleepDuration.Seconds())
+			time.Sleep(sleepDuration)
+		}
+	}
+	elapsed := time.Since(start)
+	actualQPS := float64(i) / elapsed.Seconds()
+	log.Printf("%s ended after %v : %d calls. qps=%g", tIDStr, elapsed, i, actualQPS)
+	if (numCalls > 0) && (verbose > 0) {
+		cF.Log(tIDStr+" Function duration", 99)
+		if verbose > 2 {
+			cS.Log(tIDStr+" Sleep time", 50)
+		} else {
+			cS.Counter.Log(tIDStr + " Sleep time")
+		}
+	}
+}
+
+// ParsePercentiles extracts the percentiles from string (flag).
+func ParsePercentiles(percentiles string) ([]float64, error) {
```

Comment:
Most of the Istio components use [spf13/pflag](https://github.com/spf13/pflag) which provides [IntSliceVar](https://github.com/spf13/pflag/blob/master/int_slice.go#L95) and does most of what you need here already.